### DISCUSSION TITLE
Roadmap to major version 4.0

### DISCUSSION THREAD

**Discussion by ramunasd** [php-amqplib/php-amqplib#1123] [General]
Hi all, I'd like to start discussion about next major library version - what we want to have in it and how we will approach. So far it's only proposal and all ideas are welcome. Would be nice to hear what You think about this list.

## 3.7

- [ ] last from 3.x series
- [ ] no new features, only deprecation warnings and bugfixes

## 4.0
- [ ] All connections lazy by default
- [ ] AMQPSSLConnection create only secure connections #1067
- [ ]  consistent connection constructors - using connection config only
- [ ]  proper timeout values #715
- [ ] improved defaults #667
- [ ]  typehints for all public methods
- [ ]  drop all deprecated features and constants
- [ ]  drop supprot for AMQP protocol 0.8
- [ ]  drop support for PHP <7.4
- [ ] drop support for phpseclib v2
- [ ] Support for different heartbeat negotiation options #1215

**Comment by ramunasd** [comment]
I'm planning changes for next major release, maybe You have any ideas? @lukebakken

**Reply by lukebakken** [reply]
I'll think about it! I'm busy with the next .NET client major release (https://github.com/rabbitmq/rabbitmq-dotnet-client/milestone/53)

**Comment by ots-georgewilkins** [comment]
It appears _"AMQPSSLConnection create only secure connections"_ and #1067 have been superseded by the recent #1178 and `v3.7.0`, which deprecates `AMQPSSLConnection` entirely.

Presumably direct use of `AMQPStreamConnection` is the expected alternative to `AMQPSSLConnection` as of `v4.0`?

**Reply by ramunasd** [reply]
It is not recommended to use connection classes directly due to upcoming breaking changes. Best way is through AMQPConnectionFactory, then You don't need to worry about specific classes and alternatives.

**Reply by ots-georgewilkins** [reply]
Understood, thank you @ramunasd.
